### ISSUE TITLE
A bug for calculating attention weight.

### ISSUE THREAD

**Issue by forence** [ruotianluo/self-critical.pytorch#89]
Hi RuoTian:
    I find a bug in your code where is used to calculate attention weight.
https://github.com/ruotianluo/self-critical.pytorch/blob/e3c7c944a22829cf4d936b7f0c7412220153960e/models/AttModel.py#L550-L563
    In my opinion, assume that there are different number of object proposals,  although you use mask, `weight = F.softmax(dot, dim=1)` still contain those 'empty' proposals' weight in denominator.

**Comment by ruotianluo** [comment]
That's the reason why I need next 3 lines?

**Comment by forence** [comment]
我注意到你用后面的三行了，可是你前面算weight的时候分母里是有空区域的系数，非空区域的softmax是不是被空区域影响了？即使你后面加mask把空区域的weight屏蔽掉，非空区域保持前面的不变对吧？我的意思是不是应该把mask屏蔽放在计算softmax前面，再计算softmax？

**Comment by ruotianluo** [comment]
没关系的，你可以自己推一下~
